### PR TITLE
printf compatibility

### PR DESCRIPTION
**Pull request by samueltardieu**
This makes printf more compatible with its GNU coreutils counterpart.

- format: %c prints the first character of a string
- format: parse as many characters as possible in numbers and make the error messages more compatible with GNU coreutils

### CODE REVIEW

**Comment by samueltardieu** [comment]
## About the organization

The split between the printf application and core functionality seems weird sometimes, as error messages are emitted from the core library. Maybe it would make sense if the parsing functionalities were either brought back into the printf application or returned `Result` objects in order to be able to format error messages in the printf application specifically.

For example, printf uses `‘XXX’` quotes which AFAICT were not present in the core library; emitting the message from there (`uucore::features::format::argument`) forces to use those quotes in `uucore`.

## About fuzzing

The fuzzing now often fails because too many file descriptors are opened. I don't know if there is a descriptor leak in the fuzzing framework itself, or if this is due to the launch of the external GNU coreutils commands.

**Comment by tertsdiepraam** [comment]
Re: quotes. These probably depend on locale. We chose to follow the C locale until we have proper locale support, which uses `'`.

**Comment by samueltardieu** [comment]
> Re: quotes. These probably depend on locale. We chose to follow the C locale until we have proper locale support, which uses `'`.

Interesting. I only had set `LC_NUMERIC=C`, but by setting all I get different results in the fuzzer but not in the tests. I'll fix the quotes.

**Comment by samueltardieu** [comment]
Fixed quotes

**Comment by samueltardieu** [comment]
I've restructured the code to make names more explicit, and added uucode tests to the partial parsing function.

**Review by tertsdiepraam — COMMENTED** [review]
Hi! Thanks! I think this is mostly correct, but I think it should be implemented slightly differently. A problem with this implementation is that the base is lost. Here's GNU:
```
> printf %d 010b
printf: ‘010b’: value not completely converted
8
```

So I think this is what should happen:
1. Remove the base prefix to determine the base.
2. Get the digits from the front.
3. If anything is left, print the error.
4. Convert the digits with `from_str_radix`.

This will also prevent us from trying to parse the same value multiple times, which feels a bit strange and requires us to keep both passes in sync.

Does that all make sense?

**Comment by samueltardieu** [comment]
> Hi! Thanks! I think this is mostly correct, but I think it should be implemented slightly differently. A problem with this implementation is that the base is lost. Here's GNU:
> 
> ```
> > printf %d 010b
> printf: ‘010b’: value not completely converted
> 8
> ```

TIL I learned that coreutils inherited the "octal prefix mistake" from C.

> So I think this is what should happen:
> 
>     1. Remove the base prefix to determine the base.
>     2. Get the digits from the front.
>     3. If anything is left, print the error.
>     4. Convert the digits with `from_str_radix`.

Note that:

- The "-" prefix must be extracted first as it happens earlier.
- It has to account that some bases (dec/hex) can have a fractional part while others (bin/oct) cannot.

The code I implemented only deals with decimal numbers, as those are more complicated to parse (only one "-" optional prefix allowed, only one optional "." allowed, "-inf", "inf", "nan" — although those three aren't tied to any specific base they don't accept a base prefix).

> This will also prevent us from trying to parse the same value multiple times, which feels a bit strange and requires us to keep both passes in sync.
> 
> Does that all make sense?

Yes it does. I see two alternatives:

- Keep the parser I wrote for the decimal case. It is only called when the number could not be parsed as-is, so those situations can be considered unlikely. Handle other bases numbers differently.
- Write a parser accomodating all cases and return the parsed value as well as the rest of the string if it cannot be parsed.

The second one makes more sense to me, I'll see what I can do and set this PR as draft in the meantime.

Thanks for your feedback.

**Comment by tertsdiepraam** [comment]
> TIL I learned that coreutils inherited the "octal prefix mistake" from C.

Haha yeah and now we're stuck with it 😄 

> Write a parser accomodating all cases and return the parsed value as well as the rest of the string if it cannot be parsed.

That sounds great! Thank you!

**Comment by samueltardieu** [comment]
I have implemented a new number parser which can also cope with hexadecimal floats. I have put it in another module as to differentiate between the part which emits error messages and the one which only does some processing.

**Comment by samueltardieu** [comment]
As far as I know, the only remaining issue before we can activate the printf fuzzer (in matching stderr mode) is that double quotes are escaped when printed by uutils while they are not when printed by GNU coreutils.

**Comment by samueltardieu** [comment]
The latest push just adds a new test for the error message corresponding to a larger single character:

```bash
 $ target/debug/coreutils printf %d "'abc"
printf: warning: bc: character(s) following character constant have been ignored
97%
```

**Comment by samueltardieu** [comment]
The failing chown test seems unrelated to those changes and fails for me the same way locally with and without this patch.

File: src/uucore/src/lib/features/format/argument.rs [hunk 1]
```diff
@@ -143,3 +90,39 @@ impl<'a, T: Iterator<Item = &'a FormatArgument>> ArgumentIter<'a> for T {
...
+                    if input.starts_with('\'') {
+                        show_warning!(
+                            "{}: character(s) following character constant have been ignored",
+                            &rest,
+                        );
+                    } else {
+                        show_error!("{}: value not completely converted", input.quote());
+                    }
+                    v
+                }
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/format/argument.rs:124]: Maybe it makes sense to split these cases into two enum variants?

File: src/uucore/src/lib/features/format/num_parser.rs [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+            }
+        } else {
+            (Base::Decimal, unsigned)
+        };
+        if rest.is_empty() {
+            return Err(ParseError::NotNumeric);
+        }
+        let (mut index, mut int, mut point, mut frac, mut prec, mut ended) =
+            (0, 0u64, false, 0u64, 0, false);
+        for c in rest.chars() {
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/format/num_parser.rs:196]: A couple of things about this loop:
- It might be easier to use bytes instead of chars because we only care about the digits which are in ascii range.
- I think it might be nicer to split this loop into before and after the integral part, so encountering a `.` will break the first loop and will go to the second. The current version is a bit hard to follow with all the mutable variables that keep track of the state (though I can't find any problems with it :) )

**Review by tertsdiepraam — COMMENTED** [review]
Very nice! This is great stuff! I have some small suggestions, but this could also be merged as is in my opinion.

File: src/uucore/src/lib/features/format/argument.rs [hunk 3]
```diff
@@ -41,14 +41,7 @@ impl<'a, T: Iterator<Item = &'a FormatArgument>> ArgumentIter<'a> for T {
...
             FormatArgument::Unparsed(s) => {
-                let mut chars = s.chars();
-                let Some(c) = chars.next() else {
-                    return '\0';
-                };
-                let None = chars.next() else {
-                    return '\0';
-                };
-                c
+                s.chars().next().unwrap_or('\0')
```
[the remark this answers is not in the document]
Comment (samueltardieu) [reply src/uucore/src/lib/features/format/argument.rs]: I would prefer to "fix" it in a second phase, because you might end up breaking up a multi-bytes UTF8 sequence and we have to be careful about this if it ends up in a `str`.

File: src/uucore/src/lib/features/format/num_parser.rs (hunk 2 above)
Comment (samueltardieu) [reply src/uucore/src/lib/features/format/num_parser.rs:196]: About chars / bytes : sure, but we might encounter a non-ASCII char in the process and we have to deal with it atomically.

About separating the loops, let me check if it makes things simpler.

Thanks for the review!

File: src/uucore/src/lib/features/format/argument.rs (hunk 1 above)
Comment (tertsdiepraam) [reply src/uucore/src/lib/features/format/argument.rs:124]: That does make sense, but they could be qualified as a numeric partial match and a character partial match, which are two quite different operations. I'm mostly suggesting it because it would save us from storing but discarding `rest`. I'm happy with this though, so let's leave it as it is.

File: src/uucore/src/lib/features/format/num_parser.rs (hunk 2 above)
Comment (tertsdiepraam) [reply src/uucore/src/lib/features/format/num_parser.rs:196]: > but we might encounter a non-ASCII

I think we won't split on that if we encounter it, because it won't be a digit or a point or something else we can use. I think it should be safe.

Comment (tertsdiepraam) [reply src/uucore/src/lib/features/format/num_parser.rs:196]: To expand on this a bit. The coreutils are weird regarding utf8. The GNU versions usually do not care at all whether a string is valid. The "correct" (i.e. compatible) solution is usually to use bytes throughout and avoid using `str/String` altogether. Then for printing in error messages we do a `to_string_lossy` or something like that. It's a bit messy unfortunately.

For example, I bet that `printf` will happily accept something that starts with digits and then has some invalid utf8. I can trigger this in `fish`:
```
# GNU
> printf %d 10\xf0
printf: ‘10\360’: value not completely converted
10
# uutils
> printf %d 10\xf0
error: invalid UTF-8 was detected in one or more arguments

Usage: target/debug/coreutils printf FORMATSTRING [ARGUMENT]...
       target/debug/coreutils printf FORMAT [ARGUMENT]...
       target/debug/coreutils printf OPTION

Comment (samueltardieu) [reply src/uucore/src/lib/features/format/num_parser.rs]: @tertsdiepraam Breaking the loop in two parts has indeed decreased the complexity a lot. It also made a better structure to add inline comments. Thanks for the suggestion!

File: src/uucore/src/lib/features/format/argument.rs (hunk 1 above)
Comment (samueltardieu) [reply src/uucore/src/lib/features/format/argument.rs:124]: Note that `rest` is only a `&str`, no copy is ever made of the character data. I think the cost is insignificant.

File: src/uucore/src/lib/features/format/num_parser.rs (hunk 2 above)
Comment (samueltardieu) [reply src/uucore/src/lib/features/format/num_parser.rs]: > To expand on this a bit. The coreutils are weird regarding utf8. The GNU versions usually do not care at all whether a string is valid. The "correct" (i.e. compatible) solution is usually to use bytes throughout and avoid using `str/String` altogether. Then for printing in error messages we do a `to_string_lossy` or something like that. It's a bit messy unfortunately.

Agreed, but it comes back to the previous question: do we want to be compatible with the GNU version even when we go out of the scope of its intended usage, or is it enough to be compatible with the GNU version for its intended usage, and then to be more correct outside of this usage?

Anyways, as far as this parser is concerned, given that it receives a `&str` as input, I suggest to keep `char` which is more readable and should infer to performance penalty, and switch to `u8` if at some point you decide to manipulate `&[u8]` inside of `printf` for example. The change would be easy and feel natural then.

**Comment by samueltardieu** [comment]
Btw, are different parameters used for configuring Clippy? The windows running requires that I had a `#[allow(clippy::cognitive_complexity)]` to the parse function, while I cannot reproduce this on Linux.

File: src/uucore/src/lib/features/format/argument.rs (hunk 1 above)
Comment (tertsdiepraam) [reply src/uucore/src/lib/features/format/argument.rs:124]: Oh yeah I didn't mean a performance overhead, just an additional thing to pass around

**Comment by samueltardieu** [comment]
The failing macos/x86_64 check seems unrelated.

File: src/uucore/src/lib/features/format/num_parser.rs [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+        } else {
+            (1.0, input)
+        };
+        let prefix = rest
+            .chars()
+            .take(3)
+            .map(|c| c.to_ascii_lowercase())
+            .collect::<String>();
+        let special = match prefix.as_str() {
+            "inf" => f64::NEG_INFINITY,
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/format/num_parser.rs:144]: Why is this `NEG_INFINITY`? Wouldn't `INFINITY` make more sense? It seems to work fine because of the `copysign`, though.

File: src/uucore/src/lib/features/format/num_parser.rs [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+        }
+        .copysign(sign);
+        if rest.len() == 3 {
+            Ok(special)
+        } else {
+            Err(ParseError::PartialMatch(special, &rest[3..]))
+        }
+    }
+
+    #[allow(clippy::cognitive_complexity)]
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/format/num_parser.rs:156]: Is this still necessary?

File: src/uucore/src/lib/features/format/num_parser.rs [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+        } else {
+            n
+        }
+    }
+
+    /// Parse a number as f64
+    pub fn parse_f64(input: &str) -> Result<f64, ParseError<'_, f64>> {
+        match Self::parse(input, false) {
+            Ok(v) => Ok(v.into_f64()),
+            Err(ParseError::NotNumeric) => Self::parse_f64_special_values(input),
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/format/num_parser.rs:127]: Hmm, I wonder why codecov is complaining that these aren't tested, because they definitely are tested below. Maybe codecov is not set up to run these unit tests? Anyway, this is not something for you to figure out I suppose :)

**Review by tertsdiepraam — COMMENTED** [review]
Final few comments I think, just some super minor things. This is excellent and I want to merge it as soon as possible.

File: src/uucore/src/lib/features/format/num_parser.rs (hunk 6 above)
Comment (tertsdiepraam) [reply src/uucore/src/lib/features/format/num_parser.rs:127]: Tried to fix it here: https://github.com/uutils/coreutils/pull/5825

File: src/uucore/src/lib/features/format/num_parser.rs (hunk 5 above)
Comment (samueltardieu) [reply src/uucore/src/lib/features/format/num_parser.rs:156]: Yup, on Windows at least, see https://github.com/uutils/coreutils/pull/5783#issuecomment-1882893673

File: src/uucore/src/lib/features/format/num_parser.rs (hunk 4 above)
Comment (samueltardieu) [reply src/uucore/src/lib/features/format/num_parser.rs]: This was an oversight. Fixed.

**Comment by tertsdiepraam** [comment]
Great! Thank you!

File: src/uucore/src/lib/features/format/argument.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uucore/src/lib/features/format/argument.rs]: I turned this into an issue: https://github.com/uutils/coreutils/issues/5826

File: src/uucore/src/lib/features/format/num_parser.rs (hunk 2 above)
Comment (tertsdiepraam) [reply src/uucore/src/lib/features/format/num_parser.rs]: > do we want to be compatible with the GNU version even when we go out of the scope of its intended usage, or is it enough to be compatible with the GNU version for its intended usage, and then to be more correct outside of this usage?

So far, the answer has been that we do indeed want to be compatible, but of course the intended usage has more priority. It's difficult to tell what even is the intended usage at this point, because the coreutils have accumulated so much functionality. I think the strength of this particular project lies in the compatibility, even though I often wish I could change (i.e. "fix") the behaviour. But I'm always happy to discuss specific cases :)
